### DISCUSSION TITLE
Migrating Channels from Mirth 3.12 to Mirth 4.5

### DISCUSSION THREAD

**Discussion by GokulAbhilash** [nextgenhealthcare/connect#6294] [Q&A]
I am trying to migrate channels from an older version of Mirth Connect 3.12to a newer version 4.5. However, I am encountering errors such as,
DETAILS: TypeError: [JavaPackage SOMEPACKAGE.CLASSNAME(which Im using)] is not a function, it is an object.
As I am new to Mirth Connect, I researched and learned that I need to move the required packages or JAR files into the custom-lib folder located at \Mirth Connect\custom-lib. I have done that, but I am still facing errors when setting up the channel and sending messages.
This is how im using packages
var x = new Packages.SomeClass(); 
x.usingitscontains
Could you please provide guidance on what steps I need to take to resolve this issue or if there are additional configurations required?

**Comment by pacmano1** [comment]
Actually, best to use the Resources tab and associate jars there and then at the channel level.   If you still get the error message, Mirth is not finding your class/jar.

**Comment by GokulAbhilash** [comment]
ERROR  (postprocessor:138): Error loading class 'Packages.SomeClass': JavaException: java.lang.ClassNotFoundException: Packages.SomeClass

I have used the resources tab and added the jar files from there. Still getting the above, so it means my class/jar files are not being recognized by mirth?
And there is no issue when I'm trying to use them in channels in mirth 3.12 version.
Am I missing any step when exporting channels from mirth 3.12 version to mirth  4.5?

**Comment by pacmano1** [comment]
Not sure why you are being obtuse about the actual code, but the syntax would normally be like in https://github.com/pacmano1/Mirth-Snippets/blob/main/JWTviaJose4jlib.js (just a code template that uses outside packages)

**Comment by GokulAbhilash** [comment]
Hello Pacmano, 
Thank you for your response, I have attached the latest update of what I have done based on you suggestions.

Transformer error
ERROR MESSAGE: Error evaluating transformer
com.mirth.connect.server.MirthJavascriptTransformerException: 
CHANNEL:	Test Java Channel
CONNECTOR:	Destination 1
SCRIPT SOURCE:	TRANSFORMER
SOURCE CODE:	
101: 
102: 
103: var Greeting = Packages.JavaTestPacks.Greeting;
104: 
105: 
106: var greetingInstance = new Greeting();
107: 
108: 
109: var result = greetingInstance.sayHello("John");
110: 
LINE NUMBER:	106
DETAILS:	TypeError: [JavaPackage JavaTestPacks.Greeting] is not a function, it is object.

This is the error I'm seeing when I'm trying to run in channel. This time I have used a custom test package and its still showing the same kind of error.
```
var Greeting = Packages.JavaTestPacks.Greeting;


var greetingInstance = new Greeting();


var result = greetingInstance.sayHello("John");


logger.info("Greeting Result: " + result);


msg = result;
```

I am also attaching the jar file & channel for the reference.
[TestChannelNJar.zip](https://github.com/user-attachments/files/16806269/TestChannelNJar.zip)

**Reply by pacmano1** [reply]
Works fine for me (I did not use your channel since I am on 4.5)

![image](https://github.com/user-attachments/assets/924b2239-f247-4b78-baf5-581ea078e5d2)

![image](https://github.com/user-attachments/assets/adb28e0a-1eb1-4e40-8fe1-42beaa2ce2e0)



![image](https://github.com/user-attachments/assets/25112196-a711-4416-8710-79fdc3f16b94)

**Comment by GokulAbhilash** [comment]
I'm receiving this error, I'm using 4.5.1 version of mirth connect. 

![image](https://github.com/user-attachments/assets/8f571214-ce35-4562-92bc-b2cc51d49fa7)

I'm using wrong version of mirth connect?

**Reply by pacmano1** [reply]
See @ab-mg-23's note below.

@GokulAbhilash changed issues in this thread.

**Reply by ab-mg-23** [reply]
The error you are getting is a duplicate of Discussion #6069. You have a transformer step that is validating the version of Mirth and bombing out because the code predated 4.X being a version of Mirth.

**Reply by pacmano1** [reply]
thanks @ab-mg-23  - I sorta remembered that thread but could not find it - and the OP was flipping topics.

**Reply by pacmano1** [reply]
Looks like in the code template "MirthUtils" that I didn't import so I would not have seen it anyway.

**Comment by GokulAbhilash** [comment]
I apologize for the confusion. I was trying out the solutions you provided in the thread, but I encountered some errors and shared them in thread. 
Since I can't share the actual jars and information, I created a test jar to try out the solutions, which have caused the issue to deviate from the original problem.

This test jar file is working now, I have deleted the channel and reimported the channel without importing "MirthUtils".

**Comment by GokulAbhilash** [comment]
The detailed steps of what im doing:

1. I have installed mirth connect 4.5 and launched it with java 8
2. I have already exported the channels and java lib from the mirth 3.12 and imported java libs in custom lib. Java lib is also in Java 8
3. I have imported the channel from 3.12 to 4.5 and it asked me whether I want to automatically convert into 4.5, I have clicked yes.
![image](https://github.com/user-attachments/assets/753861b0-fb3a-4edc-b497-58213decb529)
4. I have opened the channel and deployed the channel and ran a message.

Logging : 
In the destination there is a code which is using constructors from java lib which I have imported and the error says
Error Message: ERR|InternalError: Java constructor for "sqlconnect.DB_Connect" with arguments "net.sourceforge.jtds.jdbc.JtdsConnection,org.apache.logging.log4j.core.Logger" not found. 

I dont understand if Iam missing any steps or any configs. Can you please advice me, I have been sitting on it for a week.
Thanks for patience pacmano1.

**Reply by ab-mg-23** [reply]
Mirth went from using log4j to log4j2 as of Mirth 4.1. If you are passing in the `logger` from the Javascript context that will be a log4j2 `org.apache.logging.log4j.core.Logger` not a log4j `org.apache.logging.log4j.Logger` that your Java code expects.

**Reply by GokulAbhilash** [reply]
According to this I will configure my jar file to use log4j2 instead of log4j and will let you know the outcome.
Sorry for the delay

**Comment by GokulAbhilash** [comment]
Hello, 

Thanks for your help, I am able to resolve the logger issue. However I am still facing this error "TypeError: [JavaPackage monitoring.MVE_ACK] is not a function, it is object."

Postprocessor Script error
ERROR MESSAGE: Error running postprocessor scripts
com.mirth.connect.server.MirthJavascriptTransformerException: 
SOURCE CODE:	
999: 	var processedDestination = conn;
1000: 
1001: 	var ack_Mon;
1002: 
1003: 	ack_Mon = new Packages.monitoring.MVE_ACK('2225.25.36.10', 1996);
1004: 	
1005: 	
1006: 	if (checkString(processedDestination))
1007: 	{
1008: 
LINE NUMBER:	1004
DETAILS:	TypeError: [JavaPackage monitoring.MVE_ACK] is not a function, it is object.
----------------------------
[2024-09-08 00:39:06.029]  ERROR  (postprocessor:138): 
Message ID: 2486176
Date: 20240908003906998
Error Type: |System Exception
Error Message: ERR|TypeError: [JavaPackage sqlconnect.DB_Connect] is not a function, it is object. in line number 1030.

I'm getting this error where ever I'm trying to build an instance of a class of jar file. I'm not able to identify the root cause of this. Am I doing anything wrong? I have followed steps provided by @pacmano1 of adding jars to the resources tab and channels still I see the same error. If this means my mirth is not able to find my class then how to resolve this?
